File: src/main/java/cotizaciones/app/Model/WorkOrder.java
package cotizaciones.app.Model;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Entity
@Table(name = "work_orders")
@Getter
@Setter
public class WorkOrder {
     @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, length = 30)
    private String numero;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "quote_id")
    private Quote quote;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "customer_id", nullable = false)
    private Customer customer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "asset_id")
    private Asset asset;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private WorkOrderStatus estado = WorkOrderStatus.ABIERTA;

    @Column(columnDefinition = "text")
    private String notas;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal total = BigDecimal.ZERO;

    @Column(name = "fecha_apertura", nullable = false)
    private OffsetDateTime fechaApertura;

    @Column(name = "fecha_cierre")
    private OffsetDateTime fechaCierre;

    @Column(name = "fecha_entrega")
    private OffsetDateTime fechaEntrega;

    @OneToMany(
            mappedBy = "workOrder",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @OrderBy("orden ASC")
    private List<WorkOrderItem> items = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @PrePersist
    void onCreate() {
        var now = OffsetDateTime.now();
        createdAt = now;
        updatedAt = now;
        if (fechaApertura == null) {
            fechaApertura = now;
        }
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }

    // --- Métodos de negocio ---

    public void addItem(WorkOrderItem item) {
        items.add(item);
        item.setWorkOrder(this);
    }

    public void replaceItems(List<WorkOrderItem> nuevos) {
        items.clear();
        for (WorkOrderItem item : nuevos) {
            addItem(item);
        }
    }

    public void recalcularTotal() {
        this.total = items.stream()
                .map(WorkOrderItem::getSubtotal)
                .filter(Objects::nonNull)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}

File: frontend/src/types/workOrder.ts
export type WorkOrderStatus =
  | 'ABIERTA'
  | 'EN_PROCESO'
  | 'TERMINADA'
  | 'ENTREGADA'
  | 'CANCELADA';

export const WORK_ORDER_STATUS_LABELS: Record<WorkOrderStatus, string> = {
  ABIERTA: 'Abierta',
  EN_PROCESO: 'En proceso',
  TERMINADA: 'Terminada',
  ENTREGADA: 'Entregada',
  CANCELADA: 'Cancelada',
};

export const WORK_ORDER_STATUS_COLORS: Record<WorkOrderStatus, string> = {
  ABIERTA: 'bg-gray-100 text-gray-700',
  EN_PROCESO: 'bg-yellow-100 text-yellow-800',
  TERMINADA: 'bg-blue-100 text-blue-700',
  ENTREGADA: 'bg-green-100 text-green-700',
  CANCELADA: 'bg-red-100 text-red-700',
};

export interface WorkOrderItem {
  id: number | null;
  descripcion: string;
  cantidad: number;
  precioUnitario: number;
  subtotal: number;
  orden: number;
}

export interface WorkOrderItemRequest {
  descripcion: string;
  cantidad: number;
  precioUnitario: number;
}

export interface WorkOrder {
  id: number;
  numero: string;
  quoteId: number | null;
  quoteNumero: string | null;
  customerId: number;
  customerNombre: string;
  assetId: number | null;
  assetDescripcion: string | null;
  estado: WorkOrderStatus;
  notas: string | null;
  total: number;
  fechaApertura: string;
  fechaCierre: string | null;
  fechaEntrega: string | null;
  items: WorkOrderItem[];
  createdAt: string;
  updatedAt: string;
}

export interface WorkOrderRequest {
  customerId: number;
  assetId?: number | null;
  quoteId?: number | null;
  notas?: string | null;
  items: WorkOrderItemRequest[];
}